### PR TITLE
Jk/cumulus 3940 18.5.x

### PR DESCRIPTION
**Pull request by Jkovarik**
**Summary:** Summary of changes

Addresses [CUMULUS-9340](https://bugs.earthdata.nasa.gov/browse/CUMULUS-3940)

Reviewers - please take a look at report linked in JIRA issue for more context. 

### Changed

- **CUMULUS-3940**
  - Added 'dead_letter_recovery_cpu' and 'dead_letter_recovery_memory' to `cumulus` and `archive` module configuration to allow configuration of the dead_letter_recovery_operation task definition to better allow configuration of the tool's operating environment.
  - Updated the dead letter recovery tool to utilize it's own log group "${var.prefix}-DeadLetterRecoveryEcsLogs"
  - Added `batchSize`, `concurrency` and `maxDbPool` options to /endpoints/recoverCumulusMessage (note these values are correct at time of this release only):
    - `batchSize` - specifies how many DLA objects to read from S3 and hold in memory.  Defaults to 1000.
    - `concurrency` - specifies how many messages to process at the same time.  Defaults to 30.
    - `maxDbPool` - specifies how many database connections to allow the process to utilize.    Defaults to 50, value should target (`concurrency` * expected granules per message)
  - Add API memory-constrained performance test to test minimum functionality under default+ configuration

### Fixed

- **CUMULUS-3940**
  - Updated `process-s3-dead-letter-archive` and downstream calls to pass in a esClient to  `writeRecordsFunction` and update downstream calls to utilize the client.

## PR Checklist

- [x] Update CHANGELOG
- [x] Unit tests
- [x] Ad-hoc testing - Deploy changes and test manually
- [ ] Integration tests

### CODE REVIEW

**Comment by Jkovarik** [comment]
Updating to needs review - tests have passed manually, CI is having unrelated issues.

File: packages/api/endpoints/dead-letter-archive.js [hunk 1]
```diff
@@ -11,27 +11,39 @@ async function postRecoverCumulusMessages(req, res) {
...
     payload: {
+      batchSize,
       bucket,
+      concurrency,
       path,
     },
     stackName,
     systemBucket,
-    knexConfig: process.env,
+    knexConfig: { ...process.env, maxDbPool },
```
Comment (etcart) [inline packages/api/endpoints/dead-letter-archive.js:36]: why isn't this configured as an environment variable? I note in your tests you're testing it as set in the env

File: packages/api/lambdas/process-s3-dead-letter-archive.js [hunk 2]
```diff
@@ -102,25 +110,24 @@ async function processDeadLetterArchive({
...
+      } catch (error) {
+        log.error(`Failed to write records from cumulusMessage for dead letter ${deadLetterObject.Key} due to '${error}'`);
+        allFailedKeys.push(deadLetterObject.Key);
+        log.info('Transferring unprocessed message to new archive location');
+        await transferUnprocessedMessage(deadLetterObject.Key, bucket, deadLetterMessage);
+        throw error;
       }
-    ));
+    }),
+    { concurrency });
```
Comment (etcart) [inline packages/api/lambdas/process-s3-dead-letter-archive.js:127]: def appreciate a clear limit on concurrency

File: packages/api/tests/performance/lib/test-write-granules.js [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+
+const {
+  writeGranulesFromMessage,
+} = require('../../../lib/writeRecords/write-granules');
+const { fakeFileFactory, fakeGranuleFactoryV2 } = require('../../../lib/testUtils');
+
+test.before(async (t) => {
+  // Test configuration values
+  t.context.concurrency = 100;
+  process.env.maxDbPool = 100;
```
Comment (etcart) [inline packages/api/tests/performance/lib/test-write-granules.js:55]: here maxDbPool is being set as an env variable where elsewhere it seems to be expected as an argument to a function

File: packages/api/performance_tests.sh
Comment (etcart) [inline packages/api/performance_tests.sh:1]: not that this *can't* be run as a shell script, but why can't this performance test be done purely in js, for consistency and understandability

Comment (Jkovarik) [reply packages/api/performance_tests.sh:1]: Thanks for thinking about that - this part of the PR is *entirely* intended to be a starting point.    

-----
A couple of reasons, although the approach is certainly something we can dive into, I'm not *particularly* advocating for this setup: 

1) It works  (a super weak reason, to be sure!) 
2) Most of our test scaffolding in CI and package.json is written using bash/shell 
3) As part of my attempts to profile the tests via AVA it became clear AVA itself isn't a fan of being included/directly invoked (they have an error throw in a couple of inconvenient spots if you try to hack around that), making that sort of scripting wind up being invoking node to... invoke node (via ava) which without wrap logic winds up being inefficient and YAGNI.          I'm entirely open to suggestions, particularly with respect to future think  on what we want to do to start a better pattern here
4) It involves the least amount of backtracking if we change how we're doing it

File: packages/api/tests/performance/lib/test-write-granules.js (hunk 3 above)
Comment (Jkovarik) [reply packages/api/tests/performance/lib/test-write-granules.js:55]: Both are possible, depending on the function - without knowing which function we're talking about, I can't comment on the difference, but using env is one of the accepted ways of handling DB configuration in prod/test, generically. 

 That said, it's a fair point that this is working *because* we don't provide a default as part of the test environment configuration - that's deliberate and a convention, but explicit > implicit.      I'm fine making the update to pass it directly, particularly as this likely impacts, in theory, the admin connection. 
 
 👍🏻

File: packages/api/endpoints/dead-letter-archive.js (hunk 1 above)
Comment (Jkovarik) [reply packages/api/endpoints/dead-letter-archive.js:36]: In this instance to be explicit as to where it's coming from - in non test scenarios I think we should strongly prefer not modifying process.env directly.    @cumulus/db allows configuration defaults to be set via environment variables, this line is making it clear that value is being overridden by user input.

File: packages/api/endpoints/dead-letter-archive.js [hunk 4]
```diff
@@ -11,27 +11,39 @@ async function postRecoverCumulusMessages(req, res) {
...
   const systemBucket = process.env.system_bucket;
 
-  const { bucket, path } = (req.body === undefined ? {} : req.body);
+  const {
+    bucket,
+    path,
+    batchSize = 1000,
+    concurrency = 30,
+    maxDbPool = 50,
+  } = req.body ?? {}; // TODO ZOD!  KNEEL BEFORE ZOD
```
Comment (Jkovarik) [inline packages/api/endpoints/dead-letter-archive.js:20]: 😅  Reviewers: apologies, I just realized I did not do this/missed the comment before putting up for review.       I think review is still valuable, but be aware there will be a follow on to add zod/arg parsing.

Comment (Jkovarik) [reply packages/api/endpoints/dead-letter-archive.js]: c28cf580b75

File: packages/api/endpoints/dead-letter-archive.js (hunk 1 above)
Comment (Jkovarik) [reply packages/api/endpoints/dead-letter-archive.js:69]: Got it -  In this case the maxDbPool is referring to the connection pool that's part of the Knex object in this code.   That pool allocates DB connections across concurrent requests, which in theory functionally reduces the number of open connections needed and/or handles connection retries, etc. 

This value doesn't have anything to do with the DB cluster itself, but rather the DLA recovery invocation's limit on total connections it can use.    The value here should definitely be set such that it doesn't overwhelm the DB cluster.

**Review by etcart — APPROVED** [review]
looking through this in more detail: we're adding some more intelligent parameter typing, we're bounding the processing to a limited concurrency (and associated resource limit in the DB) and *testing* that fact. all of this looks good and well though through w/r to the encountered issue

File: packages/api/endpoints/dead-letter-archive.js (hunk 1 above)
Comment (jennyhliu) [reply packages/api/endpoints/dead-letter-archive.js:69]: Where is the document about maxDbPool?
I see knex example configuration https://knexjs.org/guide/#pool `pool: { min: 0, max: 7 }`

File: packages/api/tests/performance/lib/test-write-granules.js (hunk 3 above)
Comment (jennyhliu) [reply packages/api/tests/performance/lib/test-write-granules.js:55]: Should it be `env.dbMaxPool`? 
https://github.com/nasa/cumulus/blob/v18.5.1/packages/db/src/config.ts#L200

File: tf-modules/archive/async_operation.tf [hunk 5]
```diff
@@ -45,3 +50,46 @@ resource "aws_ecs_task_definition" "async_operation" {
 ]
 EOS
 }
+
+resource "aws_ecs_task_definition" "dead_letter_recovery_operation" {
+  depends_on               = [aws_cloudwatch_log_group.async_operation]
```
Comment (jennyhliu) [inline tf-modules/archive/async_operation.tf:55]:
```suggestion
  depends_on               = [aws_cloudwatch_log_group. dead_letter_recovery]
```

File: tf-modules/archive/async_operation.tf [hunk 6]
```diff
@@ -45,3 +50,46 @@ resource "aws_ecs_task_definition" "async_operation" {
...
+  family                   = "${var.prefix}-AsyncOperationTaskDefinition"
```
Comment (jennyhliu) [inline tf-modules/archive/async_operation.tf:56]:
```suggestion
  family                   = "${var.prefix}-DeadLetterRecoveryTaskDefinition"
```

File: tf-modules/archive/async_operation.tf [hunk 7]
```diff
@@ -45,3 +50,46 @@ resource "aws_ecs_task_definition" "async_operation" {
...
+  requires_compatibilities = ["FARGATE"]
+  network_mode             = "awsvpc"
+  execution_role_arn       = var.ecs_execution_role.arn
+  task_role_arn            = var.ecs_task_role.arn
+  cpu                      = var.dead_letter_recovery_cpu
+  memory                   = var.dead_letter_recovery_memory
+  container_definitions    = <<EOS
+[
+  {
+    "name": "AsyncOperation",
```
Comment (jennyhliu) [inline tf-modules/archive/async_operation.tf:67]:
```suggestion
    "name": "DeadLetterRecovery",
```

File: tf-modules/archive/async_operation.tf [hunk 8]
```diff
@@ -45,3 +50,46 @@ resource "aws_ecs_task_definition" "async_operation" {
...
+        "value": "${var.elasticsearch_hostname}"
+      }
+    ],
+    "image": "${var.async_operation_image}",
+    "logConfiguration": {
+      "logDriver": "awslogs",
+      "options": {
+        "awslogs-group": "${aws_cloudwatch_log_group.dead_letter_recovery.name}",
+        "awslogs-region": "${data.aws_region.current.name}",
+        "awslogs-stream-prefix": "async-operation"
```
Comment (jennyhliu) [inline tf-modules/archive/async_operation.tf:89]: dead_letter_recovery if we want seperate log group prefix.

File: tf-modules/archive/async_operation.tf [hunk 9]
```diff
@@ -3,6 +3,11 @@ resource "aws_cloudwatch_log_group" "async_operation" {
   retention_in_days = lookup(var.cloudwatch_log_retention_periods, "AsyncOperationEcsLogs", var.default_log_retention_days)
   tags = var.tags
 }
+resource "aws_cloudwatch_log_group" "dead_letter_recovery" {
+  name = "${var.prefix}-DeadLetterRecoveryEcsLogs"
+  retention_in_days = lookup(var.cloudwatch_log_retention_periods, "AsyncOperationEcsLogs", var.default_log_retention_days)
```
Comment (jennyhliu) [inline tf-modules/archive/async_operation.tf:8]:
```suggestion
  retention_in_days = lookup(var.cloudwatch_log_retention_periods, "DeadLetterRecoveryEcsLogs", var.default_log_retention_days)
```

**Review by jennyhliu — CHANGES_REQUESTED** [review]

File: packages/api/tests/performance/lib/test-write-granules.js (hunk 3 above)
Comment (jennyhliu) [reply packages/api/tests/performance/lib/test-write-granules.js:55]: consider rename all maxDbPool to existing naming dbMaxPool

Comment (Jkovarik) [reply packages/api/tests/performance/lib/test-write-granules.js:55]: Excellent point.     I'm updating the logging to make this more clear and demonstrate what *knex* thinks the setting is.    Good catch, both of you.

Comment (Jkovarik) [reply packages/api/tests/performance/lib/test-write-granules.js]: cb98efaa465

File: tf-modules/archive/async_operation.tf (hunk 8 above)
Comment (Jkovarik) [reply tf-modules/archive/async_operation.tf]: Thanks for catching the inconsistency in the group names/etc.    1f7e8957b33

**Comment by Jkovarik** [comment]
@jennyhliu thanks very much for the review 🙇🏻  - I believe I addressed the outstanding concerns/comments and adjusted the docs as well.   

@etcart / @jennyhliu  Take a look and let me know what you think.

File: packages/api/lambdas/process-s3-dead-letter-archive.js [hunk 10]
```diff
@@ -87,13 +91,23 @@ async function processDeadLetterArchive({
   path = `${process.env.stackName}/dead-letter-archive/sqs/`,
   writeRecordsFunction = writeRecords,
   batchSize = 1000,
+  concurrency = 10,
```
Comment (jennyhliu) [inline packages/api/lambdas/process-s3-dead-letter-archive.js:94]: concurrency default is 30, it also set to different value 20 in another place. Should we make it consistent?
Same for dbMaxPool.

File: docs/features/dead_letter_archive.md [hunk 11]
```diff
@@ -20,6 +20,25 @@ Sending a POST request to this endpoint will trigger a Cumulus AsyncOperation th
 
 This endpoint may prove particularly useful when recovering from extended or unexpected database outage, where messages failed to process due to external outage and there is no essential malformation of each Cumulus message.
 
+### Configurable request parameters
+
+The DLA recovery endpoint takes the following configurations that help bound performance.
```
Comment (jennyhliu) [inline docs/features/dead_letter_archive.md:25]: I think we also need to update cumulus-api endpoint document.

Comment (Jkovarik) [reply docs/features/dead_letter_archive.md:25]: @jennyhliu  agree - I'll make those updates once this PR is approved.

File: packages/api/lambdas/process-s3-dead-letter-archive.js (hunk 10 above)
Comment (Jkovarik) [reply packages/api/lambdas/process-s3-dead-letter-archive.js]: 👍🏻  cbdc9653f6b - good suggestion

**Comment by Jkovarik** [comment]
@jennyhliu  I believe 33be27afdee and 7412beea243 should address the issue you were seeing in test.      Thanks for taking a look, let me know what you think! 🙇🏻

**Review by jennyhliu — APPROVED** [review]
Good work👍 thank you!
